### ISSUE TITLE
Add download functionality

### ISSUE THREAD

**Issue by bryjbrown** [fsulib/islandora_managed_access#4]
This thread is for discussing the addition of download functionality to the module. Discuss exactly what kind of use case you have that would benefit from this feature and from that we can decide how best to implement it.

**Comment by sophie-ziegler** [comment]
Thanks for following up with this functionality discussion. 

This is the use case I have in mind:
My library holds material from various Native American communities. We've been working with these communities to understand cultural sensitivity issues related to this material. 

We're being told that some of the material is not particularly sensitive, and can be publicly accessible through our digital library. 

Other material, however, depicts religious ceremonies, or other events/persons/objects that the community would not like shared widely. These objects are kept in the digital library, but behind a log-in wall. 

If individuals from the communities in question request access to these otherwise restricted items, we not only let them see the items (via Drupal roles), but we also allow (and encourage) them to download the item. We do various types of digital repatriation, and this one way in which that happens. 

If we were to use the Managed Access module, building download functionality into it would help us with this workflow. 

I hope this is clear. Please let me know if it's not. 

And thank you again for working on this and sharing!

**Comment by uconnjeustis** [comment]
Our use case is slightly different. 

We work with multiple participants as a state-wide initiative. These participants add and manage content in the repository. The vast majority of content is publicly available. Participants add content because they want their content accessible.

However, there are exceptions that come from the stipulations of the deed of gift where content needs to be restricted from view for a specific period of time and only accessible to those who have a user account or by requesting a print sent to them. For this content, we would like to users to request access by sending an email to our central service. The reason for the email is that users access content through a presentation site where the only user accounts are repository staff. The other reason is that we have over 40 participants and having users send an email to a central location allows us to get the information to the appropriate participant.

In a way, I think our participants are looking for a mix between your functionality and a "request print".

**Comment by bryjbrown** [comment]
@scottlouisziegler @uconnjeustis It seems like these are pretty different use cases, and I'll say that neither is supported by the module as-is.

@scottlouisziegler It sounds like you want a request for access that can be vetted by an administrator? I'm assuming youd need to verify that a requestor is in fact a member of the community that has access to the item. Once this person has been verified, they should become a Drupal user, yes? As for the download functionality part, I'll need more information on what you mean when you say you want to "download" an object. What kinds of cmodels/datastreams are we talking about here specifically? 

@uconnjeustis It sounds like instead of having managed access objects redirect to a user account registration page, you would prefer to have it redirect to a contact form where a user can send an email requesting a "copy" of the object. So you do NOT want users creating new accounts on the site, correct?

With both of your use cases in mind, it seems like I might be able to tweak the way managed access profiles work so that instead of working in a "create new users" mode, they can also operate in a "send a request email" mode and the human on the other end of that email can do whatever is appropriate (create a user account for them manually after vetting, or send them the information they are requesting). Would this satisfy both of your use cases, or is there more to it that I'm missing?

PS, sorry for the late response, I've been fairly ill the past few days. I'm just now catching up on my to-do list.

**Comment by uconnjeustis** [comment]
You're correct that what we're looking at here at UConn is a modification of what you've built. I've been trying to figuring out to tweak it but this one is beyond my abilities right now. And yes, having the adding functionality of being able to "create new users" or "send a request email mode" would fit our use cases here. I apologize for not being able to help you tweak this or having any programmers to help you with this. If you don't have time, we have some hours with DGI and I could ask them to modify it for us and give that back to the community. 

Again, I really appreciate the time you're taking to look at our requests and help out.

Thanks.

**Comment by bryjbrown** [comment]
@uconnjeustis Good to know. I still have some tweaks left to make to the module before we can start using it in production early next year (some bug fixes and adding captcha to the registration form), but I'll see if I can work in a proof of concept for what were talking about during that period. 👍

**Comment by uconnjeustis** [comment]
Thanks @bryjbrown.

**Comment by sophie-ziegler** [comment]
@bryjbrown, sorry for the delayed response. You are correct, there will need to be some administrator to clear the user (to authorize that they are part of a community in question). After that, they'll need to be a Drupal user. They can then download whatever is available from their community. Specifically, we usually offer Video, Audio, Large Image and PDF content, and all datastreams (OBJ and all derivatives).
